### PR TITLE
Squeeze tensor while logging

### LINKED ISSUE #14446: Device error in ModelCheckpoint when checking monitored metric improvement after resuming training
## 🐛 Bug

A device error is raised in ModelCheckpoint callback when checking if the current epoch model is better and should be saved. 

This error is happening if training on GPU and after resuming the training from a checkpoint. Under these conditions, current metric is on GPU and the best model metric inside the callback is on CPU. 

This raises the error when the logged metric that is being monitored is a single element tensor with ndim=1, e.g torch.tensor([1.0]). If logged metric has ndim = 0, e.g torch.tensor(1.0), there is no error even thought the tensors are in different devices.

```  
File "/home/miguel/envs/roof_elements/lib/python3.8/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 495, in check_monitor_top_k
    should_update_best_and_save = monitor_op(current, self.best_k_models[self.kth_best_model_path])
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!
```
### To Reproduce

Run this code a first time and stop it after some epochs have been completed. Then uncomment the line in the main block to resume from the saved checkpoint and run again.

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import ModelCheckpoint


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum().unsqueeze(0)   # add dimension 0 to reproduce the issue
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum().unsqueeze(0)  # add dimension 0 to reproduce the issue
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run(resume_ckpt=None):

    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        callbacks=ModelCheckpoint(monitor='valid_loss', mode='min', filename='bestLoss'),
        accelerator='gpu',
        logger=None,
        default_root_dir=os.getcwd(),
        max_epochs=1000)

    # fit
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data, ckpt_path=resume_ckpt)


if __name__ == "__main__":
    resume_ckpt = None

    ## Uncomment next line in a second execution to resume from the saved checkpoint and reproduce the issue
    # resume_ckpt = os.path.join(os.getcwd(), 'checkpoints', 'bestLoss.ckpt')

    run(resume_ckpt=resume_ckpt)
```

### Expected behavior
Check if the current model is to be saved without rising the Runtime error due to the metric tensors being in different devices

### Environment
<details>
  <summary>Details</summary>

        * CUDA:
                - GPU:
                        - Quadro T1000
                - available:         True
                - version:           10.2
        * Lightning:
                - pytorch-lightning: 1.7.3
                - torch:             1.12.1
                - torchmetrics:      0.9.3
        * Packages:
                - absl-py:           1.2.0
                - aiohttp:           3.8.1
                - aiosignal:         1.2.0
                - async-timeout:     4.0.2
                - attrs:             22.1.0
                - cachetools:        5.2.0
                - certifi:           2022.6.15
                - charset-normalizer: 2.1.1
                - frozenlist:        1.3.1
                - fsspec:            2022.7.1
                - google-auth:       2.11.0
                - google-auth-oauthlib: 0.4.6
                - grpcio:            1.48.0
                - idna:              3.3
                - importlib-metadata: 4.12.0
                - markdown:          3.4.1
                - markupsafe:        2.1.1
                - multidict:         6.0.2
                - numpy:             1.23.2
                - oauthlib:          3.2.0
                - packaging:         21.3
                - pip:               22.2.2
                - protobuf:          3.19.4
                - pyasn1:            0.4.8
                - pyasn1-modules:    0.2.8
                - pydeprecate:       0.3.2
                - pyparsing:         3.0.9
                - pytorch-lightning: 1.7.3
                - pyyaml:            6.0
                - requests:          2.28.1
                - requests-oauthlib: 1.3.1
                - rsa:               4.9
                - setuptools:        63.3.0
                - six:               1.16.0
                - tensorboard:       2.10.0
                - tensorboard-data-server: 0.6.1
                - tensorboard-plugin-wit: 1.8.1
                - torch:             1.12.1
                - torchmetrics:      0.9.3
                - tqdm:              4.64.0
                - typing-extensions: 4.3.0
                - urllib3:           1.26.12
                - werkzeug:          2.2.2
                - wheel:             0.37.1
                - yarl:              1.8.1
                - zipp:              3.8.1
        * System:
                - OS:                Linux
                - architecture:
                        - 64bit
                        - ELF
                - processor:         x86_64
                - python:            3.8.10
                - version:           #49~20.04.1-Ubuntu SMP Thu Aug 4 19:15:44 UTC 2022
</details>

### Additional context

cc @carmocca @awaelchli @ninginthecloud @jjenniferdai @rohitgr7 @akihironitta

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #14446

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/CHANGELOG.md [hunk 1]
```diff
@@ -136,6 +136,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed an issue to keep downscaling the batch size in case there hasn't been even a single successful optimal batch size with `mode="power"` ([#14372](https://github.com/Lightning-AI/lightning/pull/14372))
 
 
+- Squeezed tensor values while logging ([#14489](https://github.com/Lightning-AI/lightning/pull/14489))
```
Comment (awaelchli) [inline src/pytorch_lightning/CHANGELOG.md:139]: Is this correct:

```suggestion
- Squeezed tensor values when logging with `LightningModule.log` ([#14489](https://github.com/Lightning-AI/lightning/pull/14489))
```
I'd like to know where we are doing the squeezing, so mentioning self.log should be good?

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/core/module.py [hunk 2]
```diff
@@ -424,6 +424,7 @@ def log(
             )
 
         value = apply_to_collection(value, numbers.Number, self.__to_tensor)
+        value = apply_to_collection(value, torch.Tensor, torch.squeeze)
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:427]: Why does it not get moved to device if it's not squeezed?

Comment (rohitgr7) [reply src/pytorch_lightning/core/module.py:427]: the issue is
```py
torch.lt(torch.Tensor([1.2], device=cuda), torch.Tensor(1.2, device=cpu)) -> gives an error
torch.lt(torch.Tensor(1.2), device=cuda), torch.Tensor(1.2, device=cpu)) -> works
```
with logging, we anyways deal with scalars.
we don't need them on the same device.

Comment (carmocca) [inline src/pytorch_lightning/core/module.py:427]: I would suggest checking numel_1 before squeezing.

Also (here or in a follow-up), I would merge these 3 `apply_to_collection` calls

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/core/module.py (hunk 2 above)
Comment (rohitgr7) [reply src/pytorch_lightning/core/module.py]: done!

**Comment by dionman** [comment]
Does this remain resolved in the most recent release? I get a similarly looking error as https://github.com/Lightning-AI/lightning/issues/14446

**Pull request merged**